File: problems/0801-0900/0823_mark-diameter-endpoints-in-a-tree/solution_rerooting_dp.java
class Solution {

    public String markDiameterEnds(int n, int[][] edges) {
        int[] deg = new int[n];
        for (int[] e : edges) {
            deg[e[0]]++;
            deg[e[1]]++;
        }
        int[][] adj = new int[n][];
        for (int i = 0; i < n; i++) {
            adj[i] = new int[deg[i]];
        }
        int[] fill = new int[n];
        for (int[] e : edges) {
            adj[e[0]][fill[e[0]]++] = e[1];
            adj[e[1]][fill[e[1]]++] = e[0];
        }

        // Root at node 0 and sweep once for a BFS order plus parents: children
        // always sit after their parent in the order, and both passes lean on it.
        int[] parent = new int[n];
        java.util.Arrays.fill(parent, -1);
        int[] order = new int[n];
        int count = 0;
        order[count++] = 0;
        for (int head = 0; head < count; head++) {
            int u = order[head];
            for (int v : adj[u]) {
                if (v != parent[u]) {
                    parent[v] = u;
                    order[count++] = v;
                }
            }
        }

        // Down pass, over the order reversed so each child is final before its
        // parent reads it: down[v] is the height of v's subtree. The top two
        // child chains ride along because the up pass must route around a
        // parent's best arm when the path re-enters through that arm.
        int[] down = new int[n];
        int[] second = new int[n];
        int[] bestChild = new int[n];
        java.util.Arrays.fill(bestChild, -1);
        for (int i = count - 1; i >= 0; i--) {
            int v = order[i];
            int p = parent[v];
            if (p >= 0) {
                int chain = down[v] + 1;
                if (chain > down[p]) {
                    second[p] = down[p];
                    down[p] = chain;
                    bestChild[p] = v;
                } else if (chain > second[p]) {
                    second[p] = chain;
                }
            }
        }

        // Up pass, forward over the order: up[v] is the longest path leaving
        // v's subtree through its parent, and max(down[v], up[v]) is v's
        // eccentricity. A sibling arm stands in for the parent's best arm
        // exactly when v owns that arm, which is why second was kept.
        int[] up = new int[n];
        int diameter = 0;
        for (int i = 0; i < count; i++) {
            int v = order[i];
            int p = parent[v];
            if (p >= 0) {
                int arm = v == bestChild[p] ? second[p] : down[p];
                up[v] = Math.max(up[p], arm) + 1;
            }
            diameter = Math.max(diameter, Math.max(down[v], up[v]));
        }

        // A node terminates a diameter exactly when its eccentricity equals
        // the tree's widest path, so compare and print.
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < n; i++) {
            sb.append(Math.max(down[i], up[i]) == diameter ? '1' : '0');
        }
        return sb.toString();
    }
}

File: problems-originals/3701-3800/3787_find-diameter-endpoints-of-a-tree/solution_rerooting_dp.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {string}
 */
var findSpecialNodes = function (n, edges) {
    const adj = Array.from({ length: n }, () => []);
    for (const [a, b] of edges) {
        adj[a].push(b);
        adj[b].push(a);
    }

    // Root at node 0 and sweep once for a BFS order plus parents: children
    // always sit after their parent in the order, and both passes lean on it.
    const parent = new Array(n).fill(-1);
    const order = [0];
    for (let head = 0; head < order.length; head++) {
        const u = order[head];
        for (const v of adj[u]) {
            if (v !== parent[u]) {
                parent[v] = u;
                order.push(v);
            }
        }
    }

    // Down pass, over the order reversed so each child is final before its
    // parent reads it: down[v] is the height of v's subtree. The top two
    // child chains ride along because the up pass must route around a
    // parent's best arm when the path re-enters through that arm.
    const down = new Array(n).fill(0);
    const second = new Array(n).fill(0);
    const bestChild = new Array(n).fill(-1);
    for (let i = order.length - 1; i >= 0; i--) {
        const v = order[i];
        const p = parent[v];
        if (p >= 0) {
            const chain = down[v] + 1;
            if (chain > down[p]) {
                second[p] = down[p];
                down[p] = chain;
                bestChild[p] = v;
            } else if (chain > second[p]) {
                second[p] = chain;
            }
        }
    }

    // Up pass, forward over the order: up[v] is the longest path leaving
    // v's subtree through its parent, and max(down[v], up[v]) is v's
    // eccentricity. A sibling arm stands in for the parent's best arm
    // exactly when v owns that arm, which is why second was kept.
    const up = new Array(n).fill(0);
    let diameter = 0;
    for (const v of order) {
        const p = parent[v];
        if (p >= 0) {
            const arm = v === bestChild[p] ? second[p] : down[p];
            up[v] = Math.max(up[p], arm) + 1;
        }
        diameter = Math.max(diameter, down[v], up[v]);
    }

    // A node terminates a diameter exactly when its eccentricity equals
    // the tree's widest path, so compare and print.
    let res = "";
    for (let i = 0; i < n; i++) {
        res += Math.max(down[i], up[i]) === diameter ? "1" : "0";
    }
    return res;
};
